File: trunk/PhotonShooter.cpp
/*
 *  PhotonShooter.cpp
 *  puls4r
 *
 *  Created by Camille on 06/03/08.
 *  Copyright 2008 __MyCompanyName__. All rights reserved.
 *
 */

#include "PhotonShooter.h"

void PhotonShooter::shoot(std::list<Light* > lights, std::list<Object * > objects){
	
	int lightNum = 0;
	
	for(std::list<Light* >::iterator iter = lights.begin(); iter != lights.end(); ++iter){
		lightNum++;
		
		int shooted = 0;
		int progress = 0;
		double x, y, z;
		
		stored = 0;
		
		while(shooted < (maxPhotons / nbLights)){
			
			int done = (int)round(100 * (shooted+1) / (maxPhotons / nbLights));
			
			if(progress != done){
				progress = done;
				std::cout << "\r\tLight (" << lightNum << "/" << lights.size() << ") : " << progress << "%";
				std::cout.flush();
			}
			
			/*
			 * Random emission from diffuse point light 
			 * As described by Henrik Wann Jensen
			 */
			do{
				x = (((double)std::rand() / (double)RAND_MAX) * 2.0) - 1.0;
				y = (((double)std::rand() / (double)RAND_MAX) * 2.0) - 1.0;
				z = (((double)std::rand() / (double)RAND_MAX) * 2.0) - 1.0;
			}while( (x*x)+(y*y)+(z*z) > 1);
			
			
			Vector* d = new Vector(x, y, z);
			Ray* ray = new Ray((*iter)->getSource(), d);
			
			//TODO replace initial energy by light->totalEnergy * color
			float energy[3] = {100.0, 100.0, 100.0};
			//---
			
			shootPhoton(ray, lights, objects, energy, false);
			shooted++;
			
			delete(d);
			delete (ray);
		}
		
		if(stored != 0){
			map->scale_photon_power(1.0 / stored);
		}
	}
}

bool PhotonShooter::russianRoulette(double d) const{
	double epsilon = ((double)std::rand() / (double)RAND_MAX); //epsilon is in [0, 1]
	
	if(epsilon < d){
		//Photon continues course
		return false;
	}
	else{
		//Absorption
		return true;
	}
}

void PhotonShooter::shootPhoton(Ray* ray, std::list<Light * > lights, std::list<Object * > objects, float energy[3], bool indirect){
	
	//TODO take object color into account for refraction
	//TODO split refraction coeff for all waves (r, g and b)
	
	Intersection* photonIntersection = getNearestIntersection(ray, objects);
	
	if(photonIntersection != 0){
		if(photonIntersection->getObject()->isReflecting()){
			
			if(russianRoulette(photonIntersection->getObject()->getKR())){
				storePhoton(photonIntersection->getPoint(), ray->getDirection(), energy);
			}
			else{
				Ray* reflected = reflectedRay(ray, photonIntersection);
				shootPhoton(reflected, lights, objects, energy, true);
				delete(reflected);
			}
		}
		
		if(photonIntersection->getObject()->isRefracting()){
			
			if(russianRoulette(photonIntersection->getObject()->getKT())){
				storePhoton(photonIntersection->getPoint(), ray->getDirection(), energy);
			}
			else{
				
				float energy2[3];
				
				//TODO refraction index -> array[3] (r,g,b)
				
				//Red
				photonIntersection->getObject()->setRefracting(true, 1.5, 1.0);
				
				energy2[0] *= energy[0] * 1.0;
				energy2[1] *= energy[1] * 0.0;
				energy2[2] *= energy[2] * 0.0;
				
				Ray* refracted = refractedRay(ray, photonIntersection, objects);
				shootPhoton(refracted, lights, objects, energy2, true);
				delete(refracted);
				
				
				//Green
				
				energy2[0] *= energy[0] * 0.0;
				energy2[1] *= energy[1] * 1.0;
				energy2[2] *= energy[2] * 0.0;
				
				photonIntersection->getObject()->setRefracting(true, 1.4, 1.0);
				
				refracted = refractedRay(ray, photonIntersection, objects);
				shootPhoton(refracted, lights, objects, energy2, true);
				delete(refracted);
				
				//Blue
				energy2[0] *= energy[0] * 0.0;
				energy2[1] *= energy[1] * 0.0;
				energy2[2] *= energy[2] * 1.0;
				
				photonIntersection->getObject()->setRefracting(true, 1.3, 1.0);
				
				refracted = refractedRay(ray, photonIntersection, objects);
				shootPhoton(refracted, lights, objects, energy2, true);
				delete(refracted);
			}
			
		}
		
		if(indirect && !photonIntersection->getObject()->isReflecting() && !photonIntersection->getObject()->isRefracting()){
			
			if(russianRoulette(0.5)){
				Color* objectColor = photonIntersection->getObject()->getEnlightment()->getColor(photonIntersection->getPoint(), photonIntersection->getNormal(), ray, lights);
				
				energy[0] *= (float)objectColor->getR();
				energy[1] *= (float)objectColor->getG();
				energy[2] *= (float)objectColor->getB();
				
				storePhoton(photonIntersection->getPoint(), ray->getDirection(), energy);
				delete(objectColor);
			}
			else{
				//TODO random reflection
			}
		}
	}
	
	delete(photonIntersection);
}

void PhotonShooter::storePhoton(Point* position, Vector* direction, float energy[3]){
	float* pos = position->toArray();
	float* dir = direction->toArray();
	
	map->store(energy, pos, dir);
	stored++;
	
	free(pos);
	free(dir);
}

Intersection* PhotonShooter::getNearestIntersection(Ray* ray, std::list<Object * > objects, double epsilon){
	
	Intersection* nearestIntersection = 0;
	
	for(std::list<Object* >::iterator iter = objects.begin(); iter != objects.end(); ++iter){
		
		Intersection* candidate = (*iter)->intersection(ray);
		
		if(candidate != 0){
			if((nearestIntersection == 0 || nearestIntersection->getT() > candidate->getT()) && (candidate->getT() > epsilon)){
				if(nearestIntersection != 0){
					delete(nearestIntersection);
				}
				
				nearestIntersection = candidate;
			}
			else{
				delete(candidate);
			}
		}
		
	}
	
	return nearestIntersection;
}

Ray* PhotonShooter::reflectedRay(Ray* ray, Intersection* intersection){
	Vector *reflectDirection = new Vector( (*ray->getDirection()) - ( ((*intersection->getNormal()) * 2.0) * ((*intersection->getNormal()) * ray->getDirection()) ) );
	reflectDirection->normalize();
	
	Ray* reflected = new Ray(intersection->getPoint(), reflectDirection);
	
	delete(reflectDirection);
	
	return reflected;
}

Ray* PhotonShooter::refractedRay(Ray* ray, Intersection* intersection, std::list<Object * > objects){
	
	Ray* refracted = 0;
	
	Ray* r1 = refractRay(ray, intersection, 1.0, intersection->getObject()->getN());
	Intersection* refractionIntersection = getNearestIntersection(r1, objects, 0.0001);
	
	if(refractionIntersection != 0){
		if(refractionIntersection->getObject() == intersection->getObject()){
			refractionIntersection->getNormal()->invert();
			Ray* r2 = refractRay(r1, refractionIntersection, intersection->getObject()->getN(), 1.0);
			refracted = new Ray(r2);
			
			delete(r2);
		}
		else{
			refracted = new Ray(r1);
		}
	}
	else{
		refracted = new Ray(r1);
	}
	
	delete(refractionIntersection);
	delete(r1);
	
	return refracted;
}

Ray* PhotonShooter::refractRay(Ray* ray, Intersection* intersection, double n1, double n2, double epsilon){
	
	double n = n2 / n1;
	double cosT = (*intersection->getNormal()) * ray->getDirection();
	double sinT2 = 1.0 - ((1.0  - cosT * cosT ) / ( n * n ));
	
	Ray* refracted;
	Vector* refractDirection;
	Point* origin;
	
	if(sinT2 > 0){
		//refraction
		refractDirection = new Vector( (( (*ray->getDirection()) - ( (*intersection->getNormal()) * cosT ) ) * (1.0 / n)) - ( (*intersection->getNormal()) * sqrt(sinT2) ));
	}
	else{
		//reflexion
		refractDirection = new Vector( (*ray->getDirection()) - ( ((*intersection->getNormal()) * 2.0) * ((*intersection->getNormal()) * ray->getDirection()) ) );
	}
	
	origin = new Point(((*refractDirection) * epsilon) + intersection->getPoint());
	refracted = new Ray(origin, refractDirection);
	
	delete(origin);
	delete(refractDirection);
	
	return refracted;
}

File: Java/puls4r/enlightment/global/photonmapping/PhotonShooter.java
/**
 * 
 */
package puls4r.enlightment.global.photonmapping;

import javax.vecmath.Color3f;
import javax.vecmath.Point3d;
import javax.vecmath.Vector3d;

import puls4r.color.Color;
import puls4r.enlightment.Enlightment;
import puls4r.math.Point3;
import puls4r.math.Vector3;
import puls4r.scene.Scene;
import puls4r.scene.objects.Light;
import puls4r.tracer.Intersection;
import puls4r.tracer.Ray;

/**
 * @author Camille
 *
 */
public class PhotonShooter {
	private static int MAX_RECURSIONS = 500;
	private static float IRRADIANCE_AREA = 0.5f;
	private static int IRRADIANCE_GATHER = 50;
	
	private int currentRecursions;
	
	private int storedDirect;
	private int storedIndirect;
	private int storedCaustic;
	
	private int maxPhotons;
	
	private PhotonMap directEnlightment;
	//private GlobalPhotonMap indirectEnlightment;
	//private GlobalPhotonMap caustics;
	
	private Scene scene;
	
	public PhotonShooter(Scene scene, int maxPhotons) {
		this.scene = scene;
		this.maxPhotons = maxPhotons;
		
		directEnlightment = new PhotonMap(maxPhotons, IRRADIANCE_GATHER, IRRADIANCE_AREA);
		//indirectEnlightment = new GlobalPhotonMap(maxPhotons, (int)(maxPhotons*IRRADIANCE_PHOTON_RATIO), IRRADIANCE_AREA);
		//caustics = new GlobalPhotonMap(maxPhotons, (int)(maxPhotons*IRRADIANCE_PHOTON_RATIO), IRRADIANCE_AREA);
		
		directEnlightment.prepare();
		//indirectEnlightment.prepare();
		//caustics.prepare();
	}
	
	private Vector3d randomDirection(){
		double x, y, z;
		
		do{
			x = Math.random();
			y = Math.random();
			z = Math.random();
		}while( (x*x)+(y*y)+(z*z) > 1);
		
		Vector3d randDir = new Vector3d(x, y, z);
		
		return randDir;
	}
	
	private boolean russianRoulette(double d){
		double e = Math.random();
		
		if(e < d){
			//Photon continues course
			return false;
		}
		else{
			//Photon absorbed
			return true;
		}
	}
	
	private void storePhoton(Point3d position, Vector3d norm, Vector3d direction, float energy[], Color3f diffColor, boolean direct, boolean indirect, boolean caustic){
		Point3 point = new Point3((float)position.x, (float)position.y, (float)position.z);
		Vector3 dir = new Vector3((float)direction.x, (float)direction.y, (float)direction.z);
		Vector3 normal = new Vector3((float)norm.x, (float)norm.y, (float)norm.z);;
		
		Color power = new Color(energy[0], energy[1], energy[2]);
		Color diffuse = new Color((float)diffColor.x, (float)diffColor.y, (float)diffColor.z);
		
		//if(direct){
			storedDirect++;
			directEnlightment.store(point, normal, dir, power, diffuse);
		//}
		
		//else if(indirect){
		//	storedIndirect++;
		//	indirectEnlightment.store(point, normal, dir, power, diffuse);
		//}
		
		//else if(caustic){
		//	storedCaustic++;
		//	caustics.store(point, normal, dir, power, diffuse);
		//}
	}
	
	private void scaleEnergy(float energy[], Color3f color){
		energy[0] *= (float)color.x;
		energy[1] *= (float)color.y;
		energy[2] *= (float)color.z;
	}
	
	private void shootPhoton(Ray ray, float energy[], boolean direct, boolean indirect, boolean caustic){
		//TODO split refraction coeff for all waves (r, g and b)
		currentRecursions++;
		
		Intersection photonIntersection = scene.getNearestIntersection(ray);
		Color3f objectColor = photonIntersection.getObject().getEnlightment().getColor(photonIntersection.getPoint(), photonIntersection.getNormal(), ray, scene.getLights());
		
		if(photonIntersection != null){
			if(photonIntersection.getObject().isReflecting()){
				
				if(russianRoulette(photonIntersection.getObject().getKR()) || (currentRecursions >= MAX_RECURSIONS)){
					storePhoton(photonIntersection.getPoint(), photonIntersection.getNormal(), ray.getDirection(), energy, objectColor, direct, indirect, caustic);
				}
				else{
					Ray reflected = scene.reflectedRay(ray, photonIntersection);
					shootPhoton(reflected, energy, false, true, false);
				}
			}
			
			if(photonIntersection.getObject().isRefracting()){
				
				if(russianRoulette(photonIntersection.getObject().getKT()) || (currentRecursions >= MAX_RECURSIONS)){
					storePhoton(photonIntersection.getPoint(), photonIntersection.getNormal(), ray.getDirection(), energy, objectColor, direct, indirect, caustic);
				}
				else{
					Ray refracted = scene.refractedRay(ray, photonIntersection);
					shootPhoton(refracted, energy, false, false, true);
				}
				
			}
			
			if(!photonIntersection.getObject().isReflecting() && !photonIntersection.getObject().isRefracting()){
				
				scaleEnergy(energy, objectColor);
				
				storePhoton(photonIntersection.getPoint(), photonIntersection.getNormal(), ray.getDirection(), energy, objectColor, direct, indirect, caustic);
				
				if(!russianRoulette(photonIntersection.getObject().getKR()) && !(currentRecursions >= MAX_RECURSIONS)){
					Vector3d randDir = randomDirection();
					Ray reflected =  new Ray(photonIntersection.getPoint(), randDir);
					shootPhoton(reflected, energy, false, true, false);
				}
			}
		}
	}
	
	public void shoot(){
		int lightNum = 0;
		
		for(Light l : scene.getLights()){
			lightNum++;
			
			int progress = 0;
			
			storedDirect = 0;
			storedIndirect = 0;
			storedCaustic = 0;
			
			while((storedDirect + storedIndirect + storedCaustic) < (maxPhotons / scene.getLights().size())){
				
				int done = (int)Math.round(100 * ((storedDirect + storedIndirect + storedCaustic)+1) / (maxPhotons / scene.getLights().size()));
				
				if(progress != done){
					progress = done;
					System.out.println("\r\tLight (" + lightNum + "/" + scene.getLights().size() + ") : " + progress + "%");
				}
				
				Vector3d randDir = randomDirection();
				Ray ray = new Ray(l.getSource(), randDir);
				
				float energy[] = {	l.getColor().x / maxPhotons, 
									l.getColor().y / maxPhotons, 
									l.getColor().z / maxPhotons};
				
				currentRecursions = 0;
				shootPhoton(ray, energy, true, false, false);
			}
		}
		
		System.out.println("--> Balancing direct photon map...");
		directEnlightment.init();
		
		System.out.println("--> Balancing indirect photon map...");
		//indirectEnlightment.init();
		
		System.out.println("--> Balancing caustic photon map...");
		//caustics.init();
	}
	
	public Color3f irradianceEstimate(Intersection intersection){
		Color3f color = new Color3f();
		Point3 point = new Point3((float)intersection.getPoint().x, (float)intersection.getPoint().y, (float)intersection.getPoint().z);
		Vector3 norm = new Vector3((float)intersection.getNormal().x, (float)intersection.getNormal().y, (float)intersection.getNormal().z);
		
		//Direct enlightment component
		Color radiance = directEnlightment.getRadiance(point, norm);
		color.set(color.x + radiance.getR(), color.y + radiance.getG(), color.z + radiance.getB());			
		
		//Indirect enlightment component
		//radiance = indirectEnlightment.getRadiance(point, norm);
		//color.set(color.x + radiance.getR(), color.y + radiance.getG(), color.z + radiance.getB());	
		
		//Caustics component
		//radiance = indirectEnlightment.getRadiance(point, norm);
		//color.set(color.x + radiance.getR(), color.y + radiance.getG(), color.z + radiance.getB());	
		
		Enlightment.normalize(color);
		
		return color;
	}
}
